File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/AchievementHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.model.AchievementEvent
import com.zeroday.model.NotificationType
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import com.zeroday.service.AchievementService
import com.zeroday.service.NotificationService
import kotlinx.serialization.json.JsonObject

class AchievementHandler(
    private val achievementService: AchievementService,
    private val notificationService: NotificationService
) : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.GET_ACHIEVEMENTS,
        RequestTypes.CLAIM_ACHIEVEMENT
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.GET_ACHIEVEMENTS -> list(ctx, playerId)
            RequestTypes.CLAIM_ACHIEVEMENT -> claim(ctx, playerId, payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported achievement request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun list(ctx: HandlerContext, playerId: String): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val visible = achievementService.visibleTo(player)
        val progressFor = player.achievementProgress
        val rows = visible.map { def ->
            val progress = progressFor[def.id]
            val percent = progress?.percent(def) ?: 0
            val unlockedAt = progress?.completedAt
            val claimed = player.unlockedAchievements.contains(def.id)
            mapOf(
                "id" to def.id,
                "name" to def.name,
                "description" to def.description,
                "category" to def.category.name,
                "target" to def.targetValue,
                "current" to (progress?.currentValue ?: 0L),
                "percent" to percent,
                "completed" to (progress?.completed == true),
                "unlocked_at" to unlockedAt,
                "reward_credits" to def.rewardCredits,
                "reward_reputation" to def.rewardReputation,
                "reward_xp" to def.rewardXp,
                "claimed" to claimed
            )
        }
        return HandlerResult.Ok(
            type = ResponseTypes.ACHIEVEMENTS,
            payload = mapOf(
                "achievements" to rows,
                "summary" to achievementService.summaryFor(player),
                "unread_achievements" to (player.unlockedAchievements.size - player.achievementProgress.count { it.value.completed })
            )
        )
    }

    private suspend fun claim(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val achievementId = try { payload.reqStr("achievement_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Achievement ID required", ErrorCategory.VALIDATION))
        }
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val def = achievementService.get(achievementId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Unknown achievement: $achievementId", ErrorCategory.NOT_FOUND))
        val progress = player.achievementProgress[achievementId]
        if (progress?.completed != true) {
            return HandlerResult.Err(ServerError.fromMessage("Achievement not yet completed", ErrorCategory.VALIDATION))
        }
        // Idempotency: re-claiming is a no-op so the client can retry.
        if (player.unlockedAchievements.contains(achievementId)) {
            return HandlerResult.Ok(
                type = ResponseTypes.ACHIEVEMENT_UNLOCKED,
                payload = mapOf(
                    "achievement_id" to achievementId,
                    "name" to def.name,
                    "rewards" to mapOf("credits" to 0L, "reputation" to 0, "xp" to 0L),
                    "already_claimed" to true
                )
            )
        }
        player.unlockedAchievements.add(achievementId)
        ctx.services.playerService.addCredits(playerId, def.rewardCredits)
        ctx.services.playerService.addReputation(playerId, def.rewardReputation)
        ctx.services.playerService.addExperience(playerId, def.rewardXp)
        notificationService.push(
            player = player,
            type = NotificationType.ACHIEVEMENT_UNLOCKED,
            title = "Achievement unlocked: ${def.name}",
            message = def.description,
            data = mapOf("achievement_id" to def.id)
        )
        return HandlerResult.Ok(
            type = ResponseTypes.ACHIEVEMENT_UNLOCKED,
            payload = mapOf(
                "achievement_id" to achievementId,
                "name" to def.name,
                "rewards" to mapOf(
                    "credits" to def.rewardCredits,
                    "reputation" to def.rewardReputation,
                    "xp" to def.rewardXp
                )
            )
        )
    }

    /**
     * Public helper so other services (e.g. WorldEventService) can
     * emit achievement events and have the player's progress updated
     * without going through a web socket message.
     */
    suspend fun emit(
        playerId: String,
        event: AchievementEvent,
        delta: Long = 1L,
        services: com.zeroday.handler.ServiceRegistry
    ): List<com.zeroday.model.AchievementUpdate> {
        val player = services.playerService.getPlayer(playerId) ?: return emptyList()
        val updates = achievementService.record(player, event, delta)
        for (update in updates.filter { it.completed }) {
            val def = achievementService.get(update.achievementId) ?: continue
            notificationService.push(
                player = player,
                type = NotificationType.ACHIEVEMENT_UNLOCKED,
                title = "Achievement unlocked: ${def.name}",
                message = def.description,
                data = mapOf("achievement_id" to def.id, "unclaimed" to "true")
            )
        }
        return updates
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/AchievementHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;
using ZeroDayMMO.Shared.Models;
using AchievementDefinition = ZeroDayMMO.Server.Services.AchievementDefinition;

namespace ZeroDayMMO.Server.Handlers;

public class AchievementHandler : IHandler
{
    private readonly IPlayerService _playerService;
    private readonly AchievementService _achievementService;
    private readonly INotificationService _notificationService;

    public string MessageType => "get_achievements";

    public AchievementHandler(
        IPlayerService playerService,
        AchievementService achievementService,
        INotificationService notificationService)
    {
        _playerService = playerService;
        _achievementService = achievementService;
        _notificationService = notificationService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return new MessageResult("error", connectionId, new { message = "Missing payload" });

        var type = payload.Value.Str("type") ?? "get_achievements";

        return type switch
        {
            "get_achievements" => await List(connectionId),
            "claim_achievement" => await Claim(connectionId, payload.Value),
            _ => new MessageResult("error", connectionId, new { message = $"Unsupported achievement request: {type}" })
        };
    }

    private async Task<IActionResult> List(string connectionId)
    {
        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var visible = _achievementService.VisibleTo(player);
        var progressFor = player.AchievementProgress;

        var rows = visible.Select(def =>
        {
            var progress = progressFor.GetValueOrDefault(def.Id);
            var percent = progress is null ? 0 :
                def.TargetValue == 0 ? 100 :
                (int)Math.Clamp(progress.CurrentValue * 100 / def.TargetValue, 0, 100);

            return new
            {
                id = def.Id,
                name = def.Name,
                description = def.Description,
                category = def.Category.ToString(),
                target = def.TargetValue,
                current = progress?.CurrentValue ?? 0L,
                percent,
                completed = progress?.Completed == true,
                unlocked_at = progress?.CompletedAt,
                reward_credits = def.RewardCredits,
                reward_reputation = def.RewardReputation,
                reward_xp = def.RewardXp,
                claimed = player.UnlockedAchievements.Contains(def.Id)
            };
        }).ToList();

        var unread = player.AchievementProgress.Count(kvp => kvp.Value.Completed) - player.UnlockedAchievements.Count;

        return new MessageResult("achievements", connectionId, new
        {
            achievements = rows,
            summary = _achievementService.SummaryFor(player),
            unread_achievements = Math.Max(0, unread)
        });
    }

    private async Task<IActionResult> Claim(string connectionId, JsonElement payload)
    {
        string achievementId;
        try
        {
            achievementId = payload.ReqStr("achievement_id");
        }
        catch
        {
            return new MessageResult("error", connectionId, new { message = "Achievement ID required" });
        }

        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var def = _achievementService.Get(achievementId);
        if (def is null)
            return new MessageResult("error", connectionId, new { message = $"Unknown achievement: {achievementId}" });

        var progress = player.AchievementProgress.GetValueOrDefault(achievementId);
        if (progress?.Completed != true)
            return new MessageResult("error", connectionId, new { message = "Achievement not yet completed" });

        if (player.UnlockedAchievements.Contains(achievementId))
        {
            return new MessageResult("achievement_unlocked", connectionId, new
            {
                achievement_id = achievementId,
                name = def.Name,
                rewards = new { credits = 0L, reputation = 0, xp = 0L },
                already_claimed = true
            });
        }

        player.UnlockedAchievements.Add(achievementId);

        await _playerService.AddCreditsAsync(connectionId, def.RewardCredits);
        await _playerService.AddReputationAsync(connectionId, def.RewardReputation);
        await _playerService.AddExperienceAsync(connectionId, def.RewardXp);

        _notificationService.AddNotification(connectionId, new Notification
        {
            Type = NotificationType.ACHIEVEMENT_UNLOCKED,
            Title = $"Achievement unlocked: {def.Name}",
            Message = def.Description,
            Data = new Dictionary<string, string> { { "achievement_id", def.Id } }
        });

        return new MessageResult("achievement_unlocked", connectionId, new
        {
            achievement_id = achievementId,
            name = def.Name,
            rewards = new
            {
                credits = def.RewardCredits,
                reputation = def.RewardReputation,
                xp = def.RewardXp
            }
        });
    }

    public async Task<List<ZeroDayMMO.Server.Services.AchievementUpdate>> Emit(
        string playerId,
        AchievementEvent eventType,
        long delta = 1)
    {
        var player = _playerService.GetPlayer(playerId);
        if (player is null)
            return new List<ZeroDayMMO.Server.Services.AchievementUpdate>();

        var updates = _achievementService.Record(player, eventType, delta);

        foreach (var update in updates.Where(u => u.Completed))
        {
                                var def = _achievementService.Get(update.AchievementId) as ZeroDayMMO.Server.Services.AchievementDefinition;
            if (def is null) continue;

            _notificationService.AddNotification(playerId, new Notification
            {
                Type = NotificationType.ACHIEVEMENT_UNLOCKED,
                Title = $"Achievement unlocked: {def.Name}",
                Message = def.Description,
                Data = new Dictionary<string, string>
                {
                    { "achievement_id", def.Id },
                    { "unclaimed", "true" }
                }
            });
        }

        return updates;
    }
}
